**same as in lab**

In [41]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: - 
  - anaconda/osx-64::ca-certificates-2019.8.28-0, anaconda/osx-64::openssl-1.1.1d-h1de35cc_2
  - anaconda/osx-64::openssl-1.1.1d-h1de35cc_2, defaults/osx-64::ca-certificates-2019.8.28-0
  - anaconda/osx-64::ca-certificates-2019.8.28-0, defaults/osx-64::openssl-1.1.1d-h1de35cc_2
  - defaults/osx-64::ca-certificates-2019.8.28-0, defaults/osx-64::openssl-1.1.1d-h1de35ccdone

# All requested packages already installed.

Solving environment: \ 
  - anaconda/osx-64::ca-certificates-2019.8.28-0, anaconda/osx-64::openssl-1.1.1d-h1de35cc_2
  - anaconda/osx-64::ca-certificates-2019.8.28-0, defaults/osx-64::openssl-1.1.1d-h1de35cc_2
  - anaconda/osx-64::openssl-1.1.1d-h1de35cc_2, defaults/osx-64::ca-certificates-2019.8.28-0
  - defaults/osx-64::ca-certificates-2019.8.28-0, defaults/osx-64::openssl-1.1.1d-h1de35ccdone

# All requested packages already installed.

Libraries imported.


**get csv file from part 2**

In [3]:
link = "https://raw.githubusercontent.com/JustToSubmit/CapstoneCoursera/master/df_NewToronto.csv"
neighborhoods = pd.read_csv(link,index_col=0)

neighborhoods.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge ,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek ,Rouge Hill ,Port Union  43.784535   
2      M1E  Scarborough       Guildwood ,Morningside ,West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

**Print shpae- how many Broughs and neighborhoods**

In [4]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 11 boroughs and 103 neighborhoods.


**get Tornoto LAtitude and Longitude**

In [5]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="can_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


**Create Toronto's map and add markers to it**

In [6]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

#markers
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

**connect to Foursquare**

In [11]:
# Define Foursquare Credentials and Version
LIMIT = 100

CLIENT_ID = 'YTPZHFTV4SGKV2S5V0H0B2AOUWN2AKVPICLOXGY5UZ0FZQJ2' 
CLIENT_SECRET = 'BU2BPZSXY2ZARQPJGSNCQCRJYV305WWHCTM5EHE5KIRMTENN'
VERSION = '20180604' 

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: YTPZHFTV4SGKV2S5V0H0B2AOUWN2AKVPICLOXGY5UZ0FZQJ2
CLIENT_SECRET:BU2BPZSXY2ZARQPJGSNCQCRJYV305WWHCTM5EHE5KIRMTENN


**explore all Toronto's neighbohoods**

In [12]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

**create new data frame for Toronto's venues**

In [13]:
toronto_venues = getNearbyVenues(names=neighborhoods['Neighbourhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude'],
                                  )

Rouge ,Malvern
Highland Creek ,Rouge Hill ,Port Union
Guildwood ,Morningside ,West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park ,Ionview ,Kennedy Park
Clairlea ,Golden Mile ,Oakridge
Cliffcrest ,Cliffside ,Scarborough Village West
Birch Cliff ,Cliffside West
Dorset Park ,Scarborough Town Centre ,Wexford Heights
Maryvale ,Wexford
Agincourt
Clarks Corners ,Sullivan ,Tam O'Shanter
Agincourt North ,L'Amoreaux East ,Milliken ,Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview ,Henry Farm ,Oriole
Bayview Village
Silver Hills ,York Mills
Newtonbrook ,Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park ,Don Mills South
Bathurst Manor ,Downsview North ,Wilson Heights
Northwood Park ,York University
CFB Toronto ,Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens ,Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West ,

**see first rows for toronto_venues**

In [14]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  \
0                          Rouge ,Malvern              43.806686   
1                          Rouge ,Malvern              43.806686   
2  Highland Creek ,Rouge Hill ,Port Union              43.784535   
3  Highland Creek ,Rouge Hill ,Port Union              43.784535   
4  Highland Creek ,Rouge Hill ,Port Union              43.784535   

   Neighborhood Longitude                      Venue  Venue Latitude  \
0              -79.194353                    Wendy's       43.807448   
1              -79.194353      Interprovincial Group       43.805630   
2              -79.160497     Chris Effects Painting       43.784343   
3              -79.160497      Royal Canadian Legion       43.782533   
4              -79.160497  Affordable Toronto Movers       43.787919   

   Venue Longitude              Venue Category  
0       -79.199056        Fast Food Restaurant  
1       -79.200378                  Print Shop  
2       -79.163742  Construction & Landscaping  
3       -79.163085                         Bar  
4       -79.162977               Moving Target

**toronto_venues shape**

In [15]:
toronto_venues.shape

(2213, 7)

**Find how many venues we have in each neighborhood**

In [16]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide ,King ,Richmond                                              100   
Agincourt                                                               5   
Agincourt North ,L'Amoreaux East ,Milliken ,Ste...                      2   
Albion Gardens ,Beaumond Heights ,Humbergate ,J...                     10   
Alderwood ,Long Branch                                                  9   
Bathurst Manor ,Downsview North ,Wilson Heights                        21   
Bayview Village                                                         4   
Bedford Park ,Lawrence Manor East                                      25   
Berczy Park                                                            55   
Birch Cliff ,Cliffside West                                             4   
Bloordale Gardens ,Eringate ,Markland Wood ,Old...                      9   
Brockton ,Exhibition Place ,Parkdale Village                           24   
Business Reply Mail Processing Centre 969 Eastern                      17   
CFB Toronto ,Downsview East                                             4   
CN Tower ,Bathurst Quay ,Island airport ,Harbou...                     16   
Cabbagetown ,St. James Town                                            42   
Caledonia-Fairbanks                                                     5   
Canada Post Gateway Processing Centre                                  12   
Cedarbrae                                                               8   
Central Bay Street                                                     84   
Chinatown ,Grange Park ,Kensington Market                              80   
Christie                                                               18   
Church and Wellesley                                                   81   
Clairlea ,Golden Mile ,Oakridge                                         9   
Clarks Corners ,Sullivan ,Tam O'Shanter                                13   
Cliffcrest ,Cliffside ,Scarborough Village West                         2   
Cloverdale ,Islington ,Martin Grove ,Princess G...                      1   
Commerce Court ,Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        7   
Deer Park ,Forest Hill SE ,Rathnelly ,South Hil...                     14   
Del Ray ,Keelesdale ,Mount Dennis ,Silverthorn                          4   
Design Exchange ,Toronto Dominion Centre                              100   
Don Mills North                                                         5   
Dorset Park ,Scarborough Town Centre ,Wexford H...                      5   
Dovercourt Village ,Dufferin                                           19   
Downsview ,North Park ,Upwood Park                                      4   
Downsview Central                                                       3   
Downsview Northwest                                                     4   
Downsview West                                                          4   
East Birchmount Park ,Ionview ,Kennedy Park                             4   
East Toronto                                                            3   
Emery ,Humberlea                                                        2   
Fairview ,Henry Farm ,Oriole                                           63   
First Canadian Place ,Underground city                                100   
Flemingdon Park ,Don Mills South                                       21   
Forest Hill North ,Forest Hill West                                     4   
Glencairn                                                               4   
Guildwood ,Morningside ,West Hill                                       7   
Harbord ,University of Toronto                                         36   
Harbourfront                                    

**Find out how many unique category we have in returned venues**

In [17]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 266 uniques categories.


## Preprocessing and Anlyzing df features

In [18]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    1               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0                 0                 0                   0      0   
1                 0                 0                   0      0   
2                 0                 0                   0      0   
3                 0                 0                   0      0   
4                 0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0       0           0              0          0         0   
1       0           0              0          0         0   
2       0           0              0          0         0   
3       0           0              0          0         0   
4       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0                0         0             0              0         0   
1     

In [19]:
toronto_onehot.shape

(2213, 266)

**Notice that we perform onehot encoding on the venue category column, this step added to our features. Hence, from 7 features, now we have 266 features**

**Next, we group the neighborhoods by using the mean of occurance of each category**

In [20]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                             Adelaide ,King ,Richmond     0.000000   
1                                            Agincourt     0.000000   
2    Agincourt North ,L'Amoreaux East ,Milliken ,St...     0.000000   
3    Albion Gardens ,Beaumond Heights ,Humbergate ,...     0.000000   
4                               Alderwood ,Long Branch     0.000000   
5      Bathurst Manor ,Downsview North ,Wilson Heights     0.000000   
6                                      Bayview Village     0.000000   
7                    Bedford Park ,Lawrence Manor East     0.000000   
8                                          Berczy Park     0.000000   
9                          Birch Cliff ,Cliffside West     0.000000   
10   Bloordale Gardens ,Eringate ,Markland Wood ,Ol...     0.000000   
11        Brockton ,Exhibition Place ,Parkdale Village     0.041667   
12   Business Reply Mail Processing Centre 969 Eastern     0.000000   
13                         CFB Toronto ,Downsview East     0.000000   
14   CN Tower ,Bathurst Quay ,Island airport ,Harbo...     0.000000   
15                         Cabbagetown ,St. James Town     0.000000   
16                                 Caledonia-Fairbanks     0.000000   
17               Canada Post Gateway Processing Centre     0.000000   
18                                           Cedarbrae     0.000000   
19                                  Central Bay Street     0.011905   
20           Chinatown ,Grange Park ,Kensington Market     0.000000   
21                                            Christie     0.000000   
22                                Church and Wellesley     0.012346   
23                     Clairlea ,Golden Mile ,Oakridge     0.000000   
24             Clarks Corners ,Sullivan ,Tam O'Shanter     0.000000   
25     Cliffcrest ,Cliffside ,Scarborough Village West     0.000000   
26   Cloverdale ,Islington ,Martin Grove ,Princess ...     0.000000   
27                      Commerce Court ,Victoria Hotel     0.000000   
28                                          Davisville     0.000000   
29                                    Davisville North     0.000000   
30   Deer Park ,Forest Hill SE ,Rathnelly ,South Hi...     0.000000   
31      Del Ray ,Keelesdale ,Mount Dennis ,Silverthorn     0.000000   
32            Design Exchange ,Toronto Dominion Centre     0.000000   
33                                     Don Mills North     0.000000   
34   Dorset Park ,Scarborough Town Centre ,Wexford ...     0.000000   
35                        Dovercourt Village ,Dufferin     0.000000   
36                  Downsview ,North Park ,Upwood Park     0.000000   
37                                   Downsview Central     0.000000   
38                                 Downsview Northwest     0.000000   
39                                      Downsview West     0.000000   
40         East Birchmount Park ,Ionview ,Kennedy Park     0.000000   
41                                        East Toronto     0.000000   
42                                    Emery ,Humberlea     0.000000   
43                        Fairview ,Henry Farm ,Oriole     0.000000   
44              First Canadian Place ,Underground city     0.000000   
45                    Flemingdon Park ,Don Mills South     0.000000   
46                 Forest Hill North ,Forest Hill West     0.000000   
47                                           Glencairn     0.000000   
48                   Guildwood ,Morningside ,West Hill     0.000000   
49                      Harbord ,University of Toronto     0.027778   
50                                        Harbourfront     0.021739   
51   Harbourfront East ,Toronto Islands ,Union Station     0.000000   
52                       High Park ,The Junction South     0.000000   
53              Highland Creek ,Rouge Hill ,Port Union     0.000000   
54                                   Hillcrest Village     0.000000   
55   Humber Bay ,King's Mill Park ,Kingsway Park So...     0.00000

**Check new shape**

In [22]:
toronto_grouped.shape

(101, 266)

**Print the most common venuew in each neighborhood. We wll print the first 4 only, feel free to change that number and explore more**

In [24]:
num_top_venues = 4

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide ,King ,Richmond----
         venue  freq
0  Coffee Shop  0.07
1   Steakhouse  0.04
2         Café  0.04
3          Bar  0.04


----Agincourt----
            venue  freq
0    Skating Rink   0.2
1  Breakfast Spot   0.2
2  Clothing Store   0.2
3          Lounge   0.2


----Agincourt North ,L'Amoreaux East ,Milliken ,Steeles East----
                      venue  freq
0                      Park   0.5
1                Playground   0.5
2  Mediterranean Restaurant   0.0
3         Mobile Phone Shop   0.0


----Albion Gardens ,Beaumond Heights ,Humbergate ,Jamestown ,Mount Olive ,Silverstone ,South Steeles ,Thistletown----
           venue  freq
0  Grocery Store   0.2
1       Pharmacy   0.1
2    Coffee Shop   0.1
3     Beer Store   0.1


----Alderwood ,Long Branch----
          venue  freq
0   Pizza Place  0.22
1           Gym  0.11
2  Skating Rink  0.11
3   Coffee Shop  0.11


----Bathurst Manor ,Downsview North ,Wilson Heights----
            venue  freq
0     Coffee Shop  0.10
1

**Sort the venues in descending order, then create new dataframe to store to top 10 venues**

In [26]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [27]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)



**show the top rows from the new dataframe**

In [28]:
neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide ,King ,Richmond           Coffee Shop   
1                                          Agincourt          Skating Rink   
2  Agincourt North ,L'Amoreaux East ,Milliken ,St...                  Park   
3  Albion Gardens ,Beaumond Heights ,Humbergate ,...         Grocery Store   
4                             Alderwood ,Long Branch           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Steakhouse                  Café                   Bar   
1        Clothing Store                Lounge        Breakfast Spot   
2            Playground      Doner Restaurant      Department Store   
3           Coffee Shop          Liquor Store   Fried Chicken Joint   
4                   Gym        Sandwich Place              Pharmacy   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                 Restaurant        Cosmetics Shop       Thai Restaurant   
1  Latin American Restaurant     Electronics Store   Empanada Restaurant   
2               Dessert Shop    Dim Sum Restaurant                 Diner   
3                   Pharmacy           Pizza Place            Beer Store   
4                       Pool                   Pub          Skating Rink   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Breakfast Spot          Burger Joint                 Bakery  
1  Eastern European Restaurant          Dessert Shop    Dumpling Restaurant  
2               Discount Store               Dog Run             Donut Shop  
3         Fast Food Restaurant        Sandwich Place         Discount Store  
4                  Coffee Shop           Gas Station                Dog Run

## Neighborhood Clustering

**we use k-means clustering. we set k(number of clusters to 5)

In [29]:

kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)


kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)


kmeans.labels_[0:10] 

array([0, 0, 1, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

**add clustering labels to our previous dataframe**

In [30]:

neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = neighborhoods

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

toronto_merged.head() 

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge ,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek ,Rouge Hill ,Port Union  43.784535   
2      M1E  Scarborough       Guildwood ,Morningside ,West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue  \
0 -79.194353             0.0  Fast Food Restaurant   
1 -79.160497             0.0                   Bar   
2 -79.188711             0.0     Electronics Store   
3 -79.216917             0.0           Coffee Shop   
4 -79.239476             0.0           Gas Station   

        2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Print Shop         Women's Store      Doner Restaurant   
1  Construction & Landscaping         Moving Target         Women's Store   
2              Breakfast Spot    Mexican Restaurant        Medical Center   
3           Korean Restaurant         Women's Store            Donut Shop   
4                      Bakery   Fried Chicken Joint    Athletics & Sports   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Dessert Shop    Dim Sum Restaurant                 Diner   
1      Doner Restaurant    Dim Sum Restaurant                 Diner   
2           Pizza Place          Intersection   Rental Car Location   
3    Dim Sum Restaurant                 Diner        Discount Store   
4  Caribbean Restaurant       Thai Restaurant                  Bank   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Discount Store               Dog Run             Donut Shop  
1        Discount Store               Dog Run             Donut Shop  
2             Drugstore            Donut Shop       Doner Restaurant  
3               Dog Run      Doner Restaurant    Dumpling Restaurant  
4      Hakka Restaurant             Drugstore             Donut Shop

**Cluster visualization**

In [61]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color = rainbow[int(cluster)-1]
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

SyntaxError: invalid syntax (<ipython-input-61-4ef7120c304f>, line 19)

**Ceate map**

In [48]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        fill=True,
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine Clusters

**Cluster 1**

In [62]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]



Borough  Cluster Labels      1st Most Common Venue  \
0         Scarborough             0.0       Fast Food Restaurant   
1         Scarborough             0.0                        Bar   
2         Scarborough             0.0          Electronics Store   
3         Scarborough             0.0                Coffee Shop   
4         Scarborough             0.0                Gas Station   
6         Scarborough             0.0         Chinese Restaurant   
7         Scarborough             0.0                   Bus Line   
8         Scarborough             0.0                      Motel   
9         Scarborough             0.0               Skating Rink   
10        Scarborough             0.0          Indian Restaurant   
11        Scarborough             0.0                 Smoke Shop   
12        Scarborough             0.0               Skating Rink   
13        Scarborough             0.0                Pizza Place   
15        Scarborough             0.0       Fast Food Restaurant   
17         North York             0.0                       Pool   
18         North York             0.0             Clothing Store   
19         North York             0.0         Chinese Restaurant   
21         North York             0.0                        Gym   
22         North York             0.0           Ramen Restaurant   
24         North York             0.0              Grocery Store   
26         North York             0.0                       Café   
27         North York             0.0                        Gym   
28         North York             0.0                Coffee Shop   
29         North York             0.0                Coffee Shop   
30         North York             0.0                       Park   
33         North York             0.0              Grocery Store   
34         North York             0.0                Coffee Shop   
35          East York             0.0       Fast Food Restaurant   
36          East York             0.0                       Park   
37       East Toronto             0.0                      Trail   
38          East York             0.0                Coffee Shop   
39          East York             0.0          Indian Restaurant   
41       East Toronto             0.0           Greek Restaurant   
42       East Toronto             0.0                       Park   
43       East Toronto             0.0                       Café   
44    Central Toronto             0.0                       Park   
45    Central Toronto             0.0                        Gym   
46    Central Toronto             0.0             Clothing Store   
47    Central Toronto             0.0               Dessert Shop   
49    Central Toronto             0.0                Coffee Shop   
51   Downtown Toronto             0.0                 Restaurant   
52   Downtown Toronto             0.0                Coffee Shop   
53   Downtown Toronto             0.0                Coffee Shop   
54   Downtown Toronto             0.0                Coffee Shop   
55   Downtown Toronto             0.0                Coffee Shop   
56   Downtown Toronto             0.0                Coffee Shop   
57   Downtown Toronto             0.0                Coffee Shop   
58   Downtown Toronto             0.0                Coffee Shop   
59   Downtown Toronto             0.0                Coffee Shop   
60   Downtown Toronto             0.0                Coffee Shop   
61   Downtown Toronto             0.0                Coffee Shop   
62         North York             0.0         Italian Restaurant   
63    Central Toronto             0.0                       Pool   
64    Central Toronto             0.0                      Trail   
65    Central Toronto             0.0                       Café   
66   Downtown Toronto             0.0                       Café   
67   Downtown Toronto             0.0                        Bar   
68   Downtown Toronto             0.0            Airport Service   
69

**Cluster 2**

In [63]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue  \
14       Scarborough             1.0                  Park   
23        North York             1.0                  Park   
25        North York             1.0                  Park   
31        North York             1.0                  Park   
40         East York             1.0                  Park   
48   Central Toronto             1.0                  Park   
50  Downtown Toronto             1.0                  Park   
74              York             1.0                  Park   
90         Etobicoke             1.0                  Park   
98              York             1.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
14            Playground      Doner Restaurant             Department Store   
23     Convenience Store                  Bank           Dim Sum Restaurant   
25     Food & Drink Shop      Doner Restaurant                 Dessert Shop   
31                  Bank         Shopping Mall                Grocery Store   
40     Convenience Store           Coffee Shop                 Dessert Shop   
48            Restaurant          Tennis Court                   Playground   
50            Playground                 Trail  Eastern European Restaurant   
74                Market  Fast Food Restaurant                Women's Store   
90                 River               Dog Run             Department Store   
98     Convenience Store          Dessert Shop           Dim Sum Restaurant   

   5th Most Common Venue    6th Most Common Venue 7th Most Common Venue  \
14          Dessert Shop       Dim Sum Restaurant                 Diner   
23                 Diner           Discount Store               Dog Run   
25    Dim Sum Restaurant                    Diner        Discount Store   
31               Airport       Falafel Restaurant  Ethiopian Restaurant   
40    Dim Sum Restaurant                    Diner        Discount Store   
48                 Diner            Deli / Bodega      Department Store   
50   Dumpling Restaurant                Drugstore     Electronics Store   
74  Colombian Restaurant  Comfort Food Restaurant   Empanada Restaurant   
90          Dessert Shop       Dim Sum Restaurant                 Diner   
98                 Diner           Discount Store               Dog Run   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
14        Discount Store               Dog Run                   Donut Shop  
23      Doner Restaurant            Donut Shop             Department Store  
25               Dog Run            Donut Shop                  Event Space  
31   Empanada Restaurant     Electronics Store  Eastern European Restaurant  
40               Dog Run      Doner Restaurant                   Donut Shop  
48          Dessert Shop    Dim Sum Restaurant               Discount Store  
50            Donut Shop      Doner Restaurant                  Curling Ice  
74     Electronics Store    College Rec Center  Eastern European Restaurant  
90        Discount Store      Doner Restaurant         Ethiopian Restaurant  
98      Doner Restaurant         Women's Store             Department Store

**Cluster 3**

In [64]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
20  North York             2.0             Cafeteria     Martial Arts Dojo   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
20         Deli / Bodega   Empanada Restaurant     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
20  Eastern European Restaurant   Dumpling Restaurant             Drugstore   

   9th Most Common Venue 10th Most Common Venue  
20            Donut Shop       Doner Restaurant

**CLuster 4**

In [65]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
5  Scarborough             3.0            Playground         Women's Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5      Doner Restaurant      Department Store          Dessert Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5    Dim Sum Restaurant                 Diner        Discount Store   

  9th Most Common Venue 10th Most Common Venue  
5               Dog Run             Donut Shop

**Cluster 5**

In [66]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  Cluster Labels       1st Most Common Venue  \
32  North York             4.0                  Food Truck   
91   Etobicoke             4.0                Home Service   
97  North York             4.0  Construction & Landscaping   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
32          Home Service        Baseball Field         Women's Store   
91        Baseball Field         Women's Store      Doner Restaurant   
97        Baseball Field         Women's Store            Donut Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
32      Department Store    Dim Sum Restaurant                 Diner   
91    Dim Sum Restaurant                 Diner        Discount Store   
97    Dim Sum Restaurant                 Diner        Discount Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
32        Discount Store               Dog Run       Doner Restaurant  
91               Dog Run            Donut Shop       Department Store  
97               Dog Run      Doner Restaurant              Drugstore